src/util/cert/cert_common_p11_child.c

```diff
@@ -0,0 +1,328 @@
+
+done:
+    if (ret != EAGAIN) {
+        if (ret == EOK) {
+            tevent_req_done(req);
+        } else {
+            tevent_req_error(req, ret);
+        }
+        tevent_req_post(req, ev);
+    }
+
+    return req;
+}
+
+static void p11_child_timeout(struct tevent_context *ev,
+                              struct tevent_timer *te,
+                              struct timeval tv, void *pvt)
+{
+    struct tevent_req *req = talloc_get_type(pvt, struct tevent_req);
+    struct cert_to_ssh_key_state *state =
+                             tevent_req_data(req, struct cert_to_ssh_key_state);
+
+    DEBUG(SSSDBG_CRIT_FAILURE, "Timeout reached for p11_child.\n");
+    child_handler_destroy(state->child_ctx);
+    state->child_ctx = NULL;
+    tevent_req_error(req, ERR_P11_CHILD);
+}
+
+static errno_t cert_to_ssh_key_step(struct tevent_req *req)
+{
+    struct cert_to_ssh_key_state *state = tevent_req_data(req,
+                                                  struct cert_to_ssh_key_state);
+    int ret;
+    int pipefd_from_child[2] = PIPE_INIT;
+    int pipefd_to_child[2] = PIPE_INIT;
+    pid_t child_pid;
+    struct timeval tv;
+
+    if (state-> iter >= state->cert_count) {
+        return EOK;
+    }
+
+    state->extra_args[0] = state->certs[state->iter];
+
+    ret = pipe(pipefd_from_child);
+    if (ret == -1) {
+        ret = errno;
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "pipe failed [%d][%s].\n", ret, strerror(ret));
+        goto done;
+    }
+    ret = pipe(pipefd_to_child);
+    if (ret == -1) {
+        ret = errno;
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "pipe failed [%d][%s].\n", ret, strerror(ret));
+        goto done;
+    }
+
+    child_pid = fork();
+    if (child_pid == 0) { /* child */
+        exec_child_ex(state, pipefd_to_child, pipefd_from_child, P11_CHILD_PATH,
+                      state->child_debug_fd, state->extra_args, false,
+                      STDIN_FILENO, STDOUT_FILENO);
+        /* We should never get here */
+        DEBUG(SSSDBG_CRIT_FAILURE, "BUG: Could not exec p11 child\n");
+    } else if (child_pid > 0) { /* parent */
+
+        state->io->read_from_child_fd = pipefd_from_child[0];
+        PIPE_FD_CLOSE(pipefd_from_child[1]);
+        sss_fd_nonblocking(state->io->read_from_child_fd);
+
+        state->io->write_to_child_fd = pipefd_to_child[1];
+        PIPE_FD_CLOSE(pipefd_to_child[0]);
+        sss_fd_nonblocking(state->io->write_to_child_fd);
+
+        /* Set up SIGCHLD handler */
+        ret = child_handler_setup(state->ev, child_pid, cert_to_ssh_key_done,
+                                  req, &state->child_ctx);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Could not set up child handlers [%d]: %s\n",
+                ret, sss_strerror(ret));
+            ret = ERR_P11_CHILD;
+            goto done;
+        }
+
+        /* Set up timeout handler */
+        tv = tevent_timeval_current_ofs(state->timeout, 0);
```

Comment:
Did you mean to use zero for the timeout?

Reply:
No, state->timeout is set to the value provide by the caller and tevent_timeval_current_ofs() adds this value to the current time.

Reply:
Oops, sorry, for some reason I thought this was absolute. Thank you.